File: scripts/translate_foods.js
const fs = require('fs');
const path = require('path');

const FOODS_FILE = path.join(__dirname, '../foods.json');

async function translateBatch(texts) {
  const text = texts.join('\n');
  const url = `https://translate.googleapis.com/translate_a/single?client=gtx&sl=en&tl=de&dt=t`;
  try {
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: `q=${encodeURIComponent(text)}`
    });
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    const data = await response.json();
    
    // Google translate returns an array of sentences
    let fullTranslated = "";
    if (data && data[0]) {
      for (const segment of data[0]) {
        if (segment[0]) {
          fullTranslated += segment[0];
        }
      }
    }
    
    return fullTranslated.split('\n').map(s => s.trim());
  } catch (error) {
    console.error(`Error translating batch:`, error);
    return null;
  }
}

async function main() {
  console.log('Loading foods...');
  const foodsData = fs.readFileSync(FOODS_FILE, 'utf8');
  const foods = JSON.parse(foodsData);
  
  console.log(`Loaded ${foods.length} items. Translating missing name_de...`);
  
  const BATCH_SIZE = 50;
  let toTranslateIndices = [];
  
  for (let i = 0; i < foods.length; i++) {
    if (!foods[i].name_de) {
      toTranslateIndices.push(i);
    }
  }
  
  console.log(`Found ${toTranslateIndices.length} items needing translation.`);
  
  for (let i = 0; i < toTranslateIndices.length; i += BATCH_SIZE) {
    const batchIndices = toTranslateIndices.slice(i, i + BATCH_SIZE);
    const texts = batchIndices.map(idx => foods[idx].name);
    
    console.log(`Translating batch ${i} to ${i + batchIndices.length}...`);
    const translatedTexts = await translateBatch(texts);
    
    if (translatedTexts && translatedTexts.length >= batchIndices.length) {
      for (let j = 0; j < batchIndices.length; j++) {
        foods[batchIndices[j]].name_de = translatedTexts[j];
      }
    } else {
      console.log(`Batch translation failed or length mismatch (${translatedTexts?.length} vs ${batchIndices.length}). Falling back to original names.`);
      for (let j = 0; j < batchIndices.length; j++) {
        foods[batchIndices[j]].name_de = foods[batchIndices[j]].name;
      }
    }
    
    fs.writeFileSync(FOODS_FILE, JSON.stringify(foods, null, 2));
    
    // Small delay between batches
    await new Promise(resolve => setTimeout(resolve, 500));
  }
  
  console.log(`Done translating. Total items processed: ${toTranslateIndices.length}`);
}

main().catch(console.error);

File: scripts/translate_foods.py
import json
import time
import sys
from deep_translator import GoogleTranslator

def load_foods(filepath):
    with open(filepath, 'r', encoding='utf-8') as f:
        return json.load(f)

def save_foods(filepath, data):
    with open(filepath, 'w', encoding='utf-8') as f:
        json.dump(data, f, indent=2, ensure_ascii=False)

def main():
    filepath = 'foods.json'
    try:
        foods = load_foods(filepath)
    except Exception as e:
        print(f"Error loading foods: {e}")
        return
    
    print(f"Loaded {len(foods)} foods.")
    
    translator = GoogleTranslator(source='en', target='de')
    
    batch_size = 500
    total_translated = 0
    
    for i, food in enumerate(foods):
        if 'name_de' not in food or not food['name_de']:
            try:
                food['name_de'] = translator.translate(food['name'])
                total_translated += 1
                if total_translated % 100 == 0:
                    print(f"Translated {total_translated} items... (current: {food['name_de']})")
            except Exception as e:
                print(f"Error translating item {food['id']} ({food['name']}): {e}")
                save_foods(filepath, foods)
                time.sleep(2)
                try:
                    food['name_de'] = translator.translate(food['name'])
                    total_translated += 1
                except Exception as e2:
                    print(f"Failed retry for {food['id']}: {e2}")
                    food['name_de'] = food['name'] # Fallback
        
        if i > 0 and i % batch_size == 0:
            save_foods(filepath, foods)
            print(f"Saved progress at {i} items.")
            
    save_foods(filepath, foods)
    print(f"Done! Translated {total_translated} new items.")

if __name__ == '__main__':
    main()
